src/cpp/src/sampler.hpp

```diff
@@ -84,6 +84,55 @@ std::vector<Token> log_softmax(const ov::Tensor& logits, size_t batch_idx) {
     return tokens;
 }
 
+// Return number of last tokens that match one of the stop_strings. If there's no match 0 is returned.
+int match_stop_string(Tokenizer & tokenizer, const TokenIds & generated_tokens, const std::set<std::string> & stop_strings) {
+    /*
+    For catching stop_string hit we run comparisons character-wise to catch cases where stop string 
+    overlaps with part of another token on both sides or is just a part of a single token. 
+    For every stop_string we iterate over generated tokens starting from the last one and going backwards. 
+    Every token is decoded and its characters are compared to the stop_string character at a current_position 
+    (position of a character in the stop_string counting from the last one) - at the begining position is 0.
+    When characters match we increase current_position and check if we have a full match already, if not we continue.
+    If we have already matched some characters (current_position > 0) and next character is not matching 
+    before we reach the full match, then we reset current_position to 0. 
+    */ 
+    for (auto stop_string: stop_strings) {
+        int current_position = 0;
+        int num_matched_tokens = 0; 
+        // Getting reverse iterator to check tokens starting from the last one generated and going backwards
+        auto generated_tokens_rit = generated_tokens.rbegin();
+        while (generated_tokens_rit != generated_tokens.rend()) {
+            num_matched_tokens++;
+            std::string decoded_token = tokenizer.decode(std::vector<int64_t>{*generated_tokens_rit});
```

Comment:
Nearby tokens may affects each other. For example a single emoji can be represented as a sequence of token_ids because of utf-8. You have to decode the last few tokens.
Tokenizers also tend to delete leading spaces. Stop string ` any` won't probably be ever matched because detokenization result of a single token doesn't start with space.

Reply:
Regarding your example of ` any` - the match is done across all generated tokens, so if last generated token is `any` then we still look for a ` ` in the next token. Even if the next token we look at is for example `find ` then we still find match for ` any` because we compare character by character. So if last 2 tokens are `find ` and `any` we effectively look for ` any` in a `find any` string so we will find a match.

> Nearby tokens may affects each other. For example a single emoji can be represented as a sequence of token_ids because of utf-8. You have to decode the last few tokens.

How far back should I look? Is it universal? 
I wonder if when I decode for examples tokens : [A], [B] - separately will I get different string than if I decode them together [A B]? 
If [A, B] is some UTF-8 character then separate decoding of A or B should result in some special character (<unk> ?) right? Because if A and B can be decoded separately to something different than [A, B] then decoding seems ambiguous ( if we generate tokens [X Y Z A B F D C] how would A B part look like) .
I'm not very familiar with tokenization/detokenization aspects, so I might be wrong though.

Reply:
Token_id representations usually have leading spaces and don't have trailing ones. So it's unlikely to get `find ` from a detokenizer.

> How far back should I look? Is it universal?

Three, maybe four token_ids. You can consult with @apaniukov.

> If [A, B] is some UTF-8 character then separate decoding of A or B should result in some special character (<unk> ?)

It should... but I think it would just crash for now. It was fixed recently: https://github.com/openvinotoolkit/openvino_tokenizers/pull/242

Reply:
> Token_id representations usually have leading spaces and don't have trailing ones. So it's unlikely to get find  from a detokenizer.

It's alright. Even if space is a separate token the match will still be found since we visit multiple tokens to find `stop_string` - `stop_string` can be a whole sentence theoretically.

Reply:
Any individual token doesn't contain a space. Spaces appear only for a sequence of tokens.

Reply:
Are you sure this is true for every tokenizer? 
I can even see your example with ` any` in https://huggingface.co/facebook/opt-125m/raw/main/vocab.json
And I've seen in my tests on opt-125m model that ` software` has been interpreted as a single token.

Reply:
That's what I'm saying. If ` any` is a single token then detokenize(tokenize(" any")) == "any". We even have a comment about it: https://github.com/openvinotoolkit/openvino.genai/blob/6a41498f4436cc0bdc3a77a9c6906c32177cf6e0/samples/cpp/prompt_lookup_decoding_lm/prompt_lookup_decoding_lm.cpp#L54

Reply:
> I wonder if when I decode for examples tokens : [A], [B] - separately will I get different string than if I decode them together [A B]?

Yes, this is the case. 
1) Right now, the default is to trim spaces at the beginning and end of the resulting string, but this behaviour will be changed in a future version of `transformers` (OV Tokenizers should catch that switch during the conversion).
2) The detokenized string should contain the entire `stop_string` so it can be detected. And the theoretical upper limit for that is the number of bytes in the longest `stop_string`.

> If [A, B] is some UTF-8 character then separate decoding of A or B should result in some special character (<unk> ?) right?

Detokenizer doesn't validate output by default. It will do it if the flag `--utf8_replace_mode="replace/ignore"` is passed during conversion (it is not supported by optimum, only by `convert_tokenizer` CLI tool or `convert_tokenizer` python function). So you get raw bytes from it and that is why all text streamers have at least 4 tokens buffer. Maybe this buffer can be reused for detecting `stop_strings`.

> Any individual token doesn't contain a space. Spaces appear only for a sequence of tokens.
There can be a token that represents an individual space and several special tokens that represent several space symbols (especially for code generation models). But if a token contains any other symbol, the space can only be at the beginning.

Reply:
> And the theoretical upper limit for that is the number of bytes in the longest stop_string

Does it mean that in worst case scenario (we only have very unique unicode characters in our output) every byte in UTF-8 encoded string is represented by a single token?

Reply:
Yes, in theory a model can produce `stop_string` byte-by-byte, but it is highly unlikely.

Reply:
Would it make sense to tokenize `stop_string` first? What I have in mind is for every stop string first check amount of tokens originally in the stop_string and if there's still no match (space trimming) I would continue decoding backwards in chunks of 4 tokens. Does it make sense? 

I need a way to keep number of generated tokens that match `stop_string` so they can be removed from the output when `include_stop_str_in_output` is set to `false` (which is a default value in model serving).

Reply:
> Would it make sense to tokenize stop_string first?
You can do it, but the model has more than one way to produce the same string output, so it won't work every time.